evap/evaluation/models.py

```diff
@@ -1627,14 +1633,18 @@ def send_to_user(self, user, subject_params, body_params, use_cc, additional_cc_
 
         subject = self.render_string(self.subject, subject_params)
         body = self.render_string(self.body, body_params)
+        self.html_body = body if self.html_body == '' else self.render_string(self.html_body, body_params)
+        html_params = dict({'content': self.html_body, 'subject': subject}, **body_params)
+        html_body = render_to_string('email_base.html', html_params)
```

Comment:
1. Paragraph: This step takes place immediately before sending the email. So for my understanding there is just an attribute of a class that gets changed and there is no further user interaction with the EmailTemplate object after this point. But maybe I'm missing something here.
2. Paragraph: I dont exactly understand what you mean as I cannot uniquely identify the code your `render_to_string(self.html_body)` reference references, but let me explain what has to be done before sending the html email, maybe that clarifies some things. First, the text of the input field has to be rendered so that the django template tags are resolved. That happens in line 1636. It either renders the `self.html_body` or just takes the `body` which is already rendered. In the next step the html template has to be filled with the subject and the content that just got rendered. This html template gets rendered in line 1638. The preparation for the rendering context parameters happens in 1637. One could argue that we should not override the `self.html_body` attribute but rather create a temporary variable.
3. Paragraph: I'm not sure if I discussed that with someone or if that was my subconsciuos understanding of the whole issue. Independent of that I think it would be cool to send every single email as a html email. For now, the only arguments I see are the following:
pro: Receivers of emails could wonder why some emails are just plain and some are html emails.
con: Users of the website are sending individual emails via the website and are wondering why their simple emails are wrapped into a big html email.

Reply:
1 and 2: I mixed two different thoughts here, let me try to make them clear(er).

I think my key point is: Methods should not have unexpected side effects.

The method name `send_to_user` does not imply any change of state on the object it is called on, so it should not change the state. This might work fine for the current code we have, but keep in mind any future code that might be written. The following code seems perfectly fine:
```python
# Send out emails to all participants
template = EmailTemplate.objects.get(type=PARTICIPANT_NOTIFICATION, ...)
for user in course.users:
    template.send_to_user(user, ...)
```

However, `template.html_body` is "corrupted" after the first loop iteration. So, as you said, we shouldn't overwrite the attribute but use a temporary variable.

---

Now, I first thought you wanted to achieve this behavior:
```
If we have no html_body yet:
    Set html_body to plain_text_body
    save the template
```

This would act as a kind of "migration", i.e. for every email template that we use, we lazily give it some html_body in case it doesn't have one. This brings us to point 3 here.

I dislike this idea. If we send html, we should utilize proper html in these emails, that is, we should use html lists and paragraphs and so on. Keep in mind that e.g. line breaks, that might be carefully crafted for the plain text mails, will be ignored for the rendering of html.

I agree that it would be cool to uniformly send out all emails as html emails, but imo this should be done by manually "translating" all templates.

---

In the end, I'd expect this method to behave like this:
```python
subject = self.render_string(self.subject, subject_params)
body = self.render_string(self.body, body_params)
mail = EmailMultiAlternatives(...)

if self.html_body:
    # notice mismatch here: we use the words "body" and "content". Maybe we can find a unified naming scheme?
    # Keep in mind: In the context of html, "body" is pretty strictly defined term that we're misusing here.
    rendered_content = self.render_string(self.html_body, body_params)
    wrapper_template_params = dict({'content': rendered_content, 'subject': subject}, **body_params)
    wrapped_content = render_to_string('email_base.html', wrapper_template_params)
    mail.attach_alternative(wrapped_content, "text/html")
```

Reply:
> 
> 
> 1 and 2: I mixed two different thoughts here, let me try to make them clear(er).
> 
> I think my key point is: Methods should not have unexpected side effects.
> 
> The method name `send_to_user` does not imply any change of state on the object it is called on, so it should not change the state. This might work fine for the current code we have, but keep in mind any future code that might be written. The following code seems perfectly fine:
> 
> ```python
> # Send out emails to all participants
> template = EmailTemplate.objects.get(type=PARTICIPANT_NOTIFICATION, ...)
> for user in course.users:
>     template.send_to_user(user, ...)
> ```
> 
> However, `template.html_body` is "corrupted" after the first loop iteration. So, as you said, we shouldn't overwrite the attribute but use a temporary variable.
> 
Okay, I see your point.
> Now, I first thought you wanted to achieve this behavior:
> 
> ```
> If we have no html_body yet:
>     Set html_body to plain_text_body
>     save the template
> ```
> 
> This would act as a kind of "migration", i.e. for every email template that we use, we lazily give it some html_body in case it doesn't have one. This brings us to point 3 here.
> 
> I dislike this idea. If we send html, we should utilize proper html in these emails, that is, we should use html lists and paragraphs and so on. Keep in mind that e.g. line breaks, that might be carefully crafted for the plain text mails, will be ignored for the rendering of html.
> 
Linebreaks are taken care of by the `|linebreaksbr` template filter, but that does not greatly weaken your argument 😅. So yes, I see your point.
> I agree that it would be cool to uniformly send out all emails as html emails, but imo this should be done by manually "translating" all templates.
> 
If you mean the staff email templates like participant notification, they all have a separate html body that uses html lists for example (see `test_data.json` lines 128103 ff.). So I guess we are only talking about individual emails that a user writes by himself on the website and then sends out to others.
To sum it up: If we would want to send every email as an html email, we should give the user a separate html email text field for editing the html body of individual emails. And as we already decided against that we shouldn't send every email as html email. Unless we find another solution...
I dont know how often such individual email get sent and whether they often have lists that should be html but that is probably no argument. So I guess I can live with only sending html emails if they have a "staff template" ^^.
> In the end, I'd expect this method to behave like this:
> 
> ```python
> subject = self.render_string(self.subject, subject_params)
> body = self.render_string(self.body, body_params)
> mail = EmailMultiAlternatives(...)
> 
> if self.html_body:
>     # notice mismatch here: we use the words "body" and "content". Maybe we can find a unified naming scheme?
>     # Keep in mind: In the context of html, "body" is pretty strictly defined term that we're misusing here.
>     rendered_content = self.render_string(self.html_body, body_params)
>     wrapper_template_params = dict({'content': rendered_content, 'subject': subject}, **body_params)
>     wrapped_content = render_to_string('email_base.html', wrapper_template_params)
>     mail.attach_alternative(wrapped_content, "text/html")
> ```
Looks good :)

Reply:
In a discussion with @janno42 we agreed that in most cases, we will only send the templates in the database. These will have `html_content` set. For the rare occasion that we send a custom email, where staff users only provide a `text_content`, we prefer a uniform look with the other emails, so we will take the plaintext, pass it through the `linebreaksbr` filter and then put it into the HTML email -- even though it may not be utilizing HTML properly in this case.